File: cpp/leetcode/字符串异位词查找/1.py
from typing import List

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        s_len , p_len = len(s), len(p)
        if s_len < p_len:
            return []

        p_count = [0]*26
        s_count = [0]*26
        ans = []

        for i in range(p_len):
            p_count[ord(p[i]) - ord('a')] += 1
            s_count[ord(s[i]) - ord('a')] += 1
        if p_count == s_count:
            ans.append(0)

        for i in range(s_len - p_len):
            s_count[ord(s[i]) - ord('a')] -= 1
            s_count[ord(s[i + p_len]) - ord('a')] += 1

            if p_count == s_count:
                ans.append(i + 1)
        return ans

if __name__ == "__main__":
    solution = Solution()
    s = "cbaebabacd"
    p = "abc"

    print(f"{solution.findAnagrams(s, p)}")

File: cpp/leetcode/字符串异位词查找/1.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int sLen = s.length(), pLen = p.length();
        if (sLen < pLen) {
            return {};
        }

        vector<int> sCount(26, 0);
        vector<int> pCount(26, 0);
        vector<int> ans;

        for (int i = 0; i < pLen; ++i) {
            pCount[p[i] - 'a']++;
            sCount[s[i] - 'a']++;
        }

        if (pCount == sCount) {
            ans.push_back(0);
        }

        for (int i = 0; i < sLen - pLen; ++i) {
            sCount[s[i] - 'a']--;
            sCount[s[i + pLen] - 'a']++;

            if (pCount == sCount) {
                ans.push_back(i + 1);
            }
        }
        return ans;
    }
};

int main() {
    string s = "cbaebabacd";
    string p = "abc";
    Solution solution;
    vector<int> result = solution.findAnagrams(s, p);
    cout << "[";
    for (const auto& i : result) {
        cout << i ;
        if (i < result.size() - 1) {
            cout << ", ";
        }
    }
    cout << "]" << endl;
}